File: lib-database/src/main/java/com/ustadmobile/core/db/dao/PersonGroupDao.java
package com.ustadmobile.core.db.dao;

import com.ustadmobile.core.db.UmLiveData;
import com.ustadmobile.core.db.UmProvider;
import com.ustadmobile.core.impl.UmCallback;
import com.ustadmobile.lib.database.annotation.UmDao;
import com.ustadmobile.lib.database.annotation.UmQuery;
import com.ustadmobile.lib.database.annotation.UmRepository;
import com.ustadmobile.lib.database.annotation.UmUpdate;
import com.ustadmobile.lib.db.entities.GroupWithMemberCount;
import com.ustadmobile.lib.db.entities.PersonGroup;
import com.ustadmobile.lib.db.sync.dao.SyncableDao;

import java.util.List;

@UmDao(updatePermissionCondition = RoleDao.SELECT_ACCOUNT_IS_ADMIN,
insertPermissionCondition = RoleDao.SELECT_ACCOUNT_IS_ADMIN)
@UmRepository
public abstract class PersonGroupDao implements SyncableDao<PersonGroup, PersonGroupDao> {

    @UmQuery("SELECT *, 0 AS memberCount FROM PersonGroup")
    public abstract UmProvider<GroupWithMemberCount> findAllGroups();

    @UmQuery("SELECT *, 0 AS memberCount FROM PersonGroup WHERE groupActive = 1")
    public abstract UmProvider<GroupWithMemberCount> findAllActiveGroups();

    @UmQuery("SELECT *, 0 AS memberCount FROM PersonGroup WHERE groupActive = 1")
    public abstract UmLiveData<List<GroupWithMemberCount>> findAllActiveGroupsLive();

    @UmQuery("SELECT * FROM PersonGroup WHERE groupActive = 1")
    public abstract UmLiveData<List<PersonGroup>> findAllActivePersonGroupsLive();

    @UmQuery("UPDATE PersonGroup SET groupActive = 0 WHERE groupUid = :uid")
    public abstract void inactivateGroupAsync(long uid, UmCallback<Integer> resultObject);

    @UmQuery("SELECT * FROM PersonGroup WHERE groupUid = :uid")
    public abstract PersonGroup findByUid(long uid);

    @UmQuery("SELECT * FROM PersonGroup WHERE groupUid = :uid")
    public abstract void findByUidAsync(long uid, UmCallback<PersonGroup> resultObject);

    @UmQuery("SELECT * FROM PersonGroup WHERE groupUid = :uid")
    public abstract UmLiveData<PersonGroup> findByUidLive(long uid);

    @UmUpdate
    public abstract void updateAsync(PersonGroup entity, UmCallback<Integer> resultObject);

}

File: lib-database/src/commonMain/kotlin/com/ustadmobile/core/db/dao/PersonGroupDao.kt
package com.ustadmobile.core.db.dao

import com.ustadmobile.door.annotation.DoorDao
import androidx.room.Query
import androidx.room.Update
import kotlinx.coroutines.flow.Flow
import com.ustadmobile.door.annotation.*
import com.ustadmobile.lib.db.entities.*

@Repository
@DoorDao
expect abstract class PersonGroupDao : BaseDao<PersonGroup> {

    @Query("SELECT * FROM PersonGroup WHERE groupUid = :uid")
    abstract fun findByUid(uid: Long): PersonGroup?

    @Query("SELECT * FROM PersonGroup WHERE groupUid = :uid")
    abstract suspend fun findByUidAsync(uid: Long) : PersonGroup?

    @Query("SELECT * FROM PersonGroup WHERE groupUid = :uid")
    abstract fun findByUidLive(uid: Long): Flow<PersonGroup?>


    @Update
    abstract suspend fun updateAsync(entity: PersonGroup) : Int

    @Query("""
        Select CASE
               WHEN Person.firstNames IS NOT NULL THEN Person.firstNames
               ELSE PersonGroup.groupName 
               END AS name
          FROM PersonGroup
               LEFT JOIN Person
                         ON Person.personGroupUid = PersonGroup.groupUid
         WHERE PersonGroup.groupUid = :groupUid
         LIMIT 1
    """)
    abstract suspend fun findNameByGroupUid(groupUid: Long): String?

}
